File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        
        diff = float('inf')
        nums = sorted(nums)
        
        for fx in range(len(nums) - 2):
            
            lx = fx + 1
            rx = len(nums) - 1
            
            while lx < rx:
                
                # calculate sum:
                isum = nums[fx] + nums[lx] + nums[rx]
                
                # check if absolute difference is less:
                if abs(target - isum) < abs(diff):
                    diff = target - isum
                    
                # move pointers based on current sum:
                if isum < target:
                    lx = lx + 1
                else:
                    rx = rx - 1
                    
                # if difference is zero, return value:
                if diff == 0:
                    break
                
        return target - diff

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        
        std::sort(nums.begin(), nums.end());
        
        int cx, rx, s;
        int d = INT_MAX;
        
        for(int lx=0; lx < nums.size() - 2; lx++) {
            
            cx = 1 + lx;
            rx = nums.size() - 1;
            
            while (cx < rx) {
                s = nums[lx] + nums[cx] + nums[rx];
                if (std::abs(target - s) < std::abs(d)) {
                    d = target - s;
                }
                if (target == s) {
                    return s;
                }
                if (s < target) {
                    cx = cx + 1;
                } else {
                    rx = rx - 1;
                }
            }
            
        }
        
        return target - d;
    }
};